deps/rabbitmq_auth_backend_oauth2/src/rabbit_auth_backend_oauth2.erl

```diff
@@ -542,24 +547,32 @@ token_from_context(AuthProps) ->
 %%   <<"sub">>         => <<"rabbit_client">>,
 %%   <<"zid">>         => <<"uaa">>}
 
--spec username_from(binary(), map()) -> binary() | undefined.
-username_from(ClientProvidedUsername, DecodedToken) ->
-    ClientId = uaa_jwt:client_id(DecodedToken, undefined),
-    Sub      = uaa_jwt:sub(DecodedToken, undefined),
-
-    rabbit_log:debug("Computing username from client's JWT token, client ID: '~ts', sub: '~ts'",
-                     [ClientId, Sub]),
+-spec username_from(list(), map()) -> binary() | undefined.
+username_from(PreferredUsernameClaims, DecodedToken) ->
+    UsernameClaims = append_or_return_default(PreferredUsernameClaims, ?DEFAULT_PREFERRED_USERNAME_CLAIMS),
+    ResolvedUsernameClaims = lists:filtermap(fun(Claim) -> find_claim_in_token(Claim, DecodedToken) end, UsernameClaims),
+    Username = case ResolvedUsernameClaims of
+      [ ] -> <<"unknown">>;
+      [ _One ] -> _One;
+      [ _One | _ ] -> _One
+    end,
+    rabbit_log:debug("Computing username from client's JWT token: ~ts -> ~ts ",
+      [lists:flatten(io_lib:format("~p",[ResolvedUsernameClaims])), Username]),
+    Username.
+
+append_or_return_default(ListOrBinary, Default) ->
+  case ListOrBinary of
+    VarList when is_list(VarList) -> VarList ++ Default;
+    VarBinary when is_binary(VarBinary) -> [VarBinary] ++ Default;
+    _ -> Default
+  end.
 
-    case uaa_jwt:client_id(DecodedToken, Sub) of
-        undefined ->
-            case ClientProvidedUsername of
-                undefined -> undefined;
-                <<>>      -> undefined;
-                _Other    -> ClientProvidedUsername
-            end;
-        Value     ->
-            Value
-    end.
+find_claim_in_token(Claim, Token) ->
```

Comment:
It is much more common to see `{ok, Value} | undefined` to be returned instead of `{true, Value} | false`.

Reply:
The reason for returning `(boolean) | {true, Value}` is because this function is used as a filter function in [filtermap](https://www.erlang.org/doc/man/lists.html#filtermap-2) and it expects those return values....